src/relay/transforms/merge_compiler_regions.cc

```diff
@@ -0,0 +1,352 @@
+    auto create = Downcast<RefCreate>(new_e);
+    return RefCreate(AddCompilerEnd(create->value));
+  }
+
+  Expr VisitExpr_(const RefReadNode *op) {
+    auto new_e = ExprMutator::VisitExpr_(op);
+    auto read = Downcast<RefRead>(new_e);
+    return RefRead(AddCompilerEnd(read->ref));
+  }
+
+  Expr VisitExpr_(const RefWriteNode *op) {
+    auto new_e = ExprMutator::VisitExpr_(op);
+    auto write = Downcast<RefWrite>(new_e);
+    return RefWriteNode::make(
+      AddCompilerEnd(write->ref),
+      AddCompilerEnd(write->value));
+  }
+
+ private:
+    AnnotatedRegionSet regions_;
+    const std::string target_ = "default";
+    Function func_;
+};
+
+class MergeAnnotations : public ExprMutator {
+ public:
+  explicit MergeAnnotations(AnnotatedRegionSet regions) : regions_(regions) {}
+
+  Expr VisitExpr_(const CallNode* call) final {
+    if (call->op == compiler_begin_op) {
+      if (call->args[0]->IsInstance<CallNode>()) {
+        auto arg = Downcast<Call>(call->args[0]);
+        if (arg->op == compiler_end_op) {
+          auto region1 = regions_->GetRegion(GetRef<Call>(call));
+          auto region2 = regions_->GetRegion(arg);
+          if (region1 == region2) {
+            return ExprMutator::VisitExpr(arg->args[0]);
+          }
+        }
+      }
+    }
+    return ExprMutator::VisitExpr_(call);
+  }
+
+ private:
+  AnnotatedRegionSet regions_;
+};
+
+class RegionMerger : public ExprVisitor {
+ public:
+  explicit RegionMerger(AnnotatedRegionSet regions) : regions_(regions) {}
+
+  void VisitExpr_(const CallNode* call) final {
+    if (call->op == compiler_end_op) {
+      auto region = regions_->GetRegion(GetRef<Call>(call));
+      // set the region target
+      auto compiler_attrs = call->attrs.as<CompilerAttrs>();
+      region_targets_[region->GetID()] = compiler_attrs->compiler;
+      std::vector<AnnotatedRegion> mergeable_regions;
+      // first look at the region args to determine the parent regions
+      for (const auto& arg : region->GetInputs()) {
+        // all args should be begin annotations
+        auto begin = Downcast<Call>(arg);
+        // the arguments of the begin annotations will be in the parent regions
+        auto parent_region = regions_->GetRegion(begin->args[0]);
+        // if there is no parent region, move on
+        if (!parent_region.defined()) continue;
+        // merge the parent region if it hasn't been done already
+        if (merged_regions_.find(parent_region->GetID()) == merged_regions_.end()) {
+          VisitExpr(begin->args[0]);
+        }
+        mergeable_regions.push_back(parent_region);
+      }
+      // region_restrictions_[region->id] = std::unordered_set<int>();
+      auto& region_restrictions = region_restrictions_[region->GetID()];
+      for (const auto& parent_region : mergeable_regions) {
+        auto parent_restrictions = region_restrictions_[parent_region->GetID()];
+        // add all the parent restrictions to the current region
```

Comment:
Move this comment to the top of this loop as other comments for the following statement.